### DISCUSSION TITLE
Would love to help get this rolling

### DISCUSSION THREAD

**Discussion by SCarter77** [lamemakes/pilot-drive#25] [General]
Would love to help move this forward. I'm not terribly good with programming, but I have used many arduinos in the past.  Currently running Openauto Pro but I really don't like the lack of customization. Aside from the steering wheel controls, I would have been better off to just grab an off-the-shelf android head unit from Amazon.
I did make my own PCB for 12v inputs and timer/shutdown circuit, program my own arduino micro to grab the resistance values from my steering wheel controls & push out button presses to the Openauto software.  Works great and I can control the button rate this way (like to have volume down much much faster than volume up).  My other issue is that the best way to get 12v to the GPIO is optocouplers but since they sink and Openauto doesn't offer pin down as a valid input you have to do a lot more yourself to flip the signal.  Huge oversight IMO
Let me know if I can help in any way

**Comment by lamemakes** [comment]
Hey! Thanks a ton for making this post, really nice to know there's still an interest. The goal is exactly what you outlined - eventually have the ability for deep customization on multiple levels - especially in terms of IO. Would love to see/chat a bit about the PCB you made! 

I need to buckle down and make a solid pass on the software to implement more modularity, customizability, and ease of install. Being the sole developer and with how big of an undertaking this project has proven to be I'm temporarily struggling with the motivational aspect of that, which explains the slower development here. But hoping to be able to be able to dive into it more soon! 

Would definitely love to hear more of your opinions about what OpenAuto got wrong & right though so I can continue to build this project out. Cheers!

**Reply by SCarter77** [reply]
My PCB was dirty lol.. I'm not good at smd components so everything was through hole.  Somewhat similar to yours but I didn't make it a hat, just a remote card.  It has 1k resistors from 12v to optocouplers tied to the inputs of an Arduino nano which handled flipping the outputs for openautos high level only signals via a level shifter to 3.3v.  It was initially going to also handle the steering wheel controls (07 Toyota FJ Cruiser) but the wiring started getting too heavy and GPIO limitations due to using a hifiberry hat already so I made a separate dongle with an Arduino Pro Micro to act as a USB keyboard & send keypresses for volume up/down and previous/next track.  Down side is using up a valuable USB port on the Pi.

 As far as the input flipping, I don't see why Openauto couldn't have just had an "invert" checkbox next to the GPIO pin selection, it would allow users (makers) much more freedom in design and lower the part count and board complexity.

 I was toying with the idea of switching to a 3.3v Arduino Pro Micro to get rid of the level shifter but decided that lowering the voltage for the steering wheel control detection from 5 to 3.3 would be pushing my luck with the noisy automotive electrical environment.

 OR.. use a USB Pro Micro or Leonardo for all inputs - this would require simple keypress grabs for lights on/off, volume up/down, next track/previous track, backup camera on/off etc.  That way you could move everything off the hat and have a separate module that only requires a USB cable to the Pi - removing bulk and cooling issues from a hat.  With Openauto this is not an option as it only has light toggle key so if you start up your car and the lights are already on, the board won't know whether it's toggling on or off.  

Wants:
 Wallpaper - loved making my own wallpaper but since the control buttons take up so much real estate (and many buttons that I don't want and are not removable), I can't fully enjoy it.  Would be nice to have the option of while on the main screen, the menu could push up to task bar and leave the current track playing across the bottom.  And since I'm on wants.. animated wallpaper would rock!  Even if just allowing animated gif support.  I made a chromed "FJ Cruiser" logo in Blender, if it could just move ever so little the effect would be awesome.
  Native file browser - One of my reasons for wanting the Pi based solution is to get all my media into it and not rely on thumbdrives.  There should be a simple way to copy from USB or a network share to internal storage.  But with Openauto, as soon as I used the network for connecting to a phone (by accident as I can't remove the button), it overrode my network connection to my home wifi.
  Dashcam - why have this powerful little computer in your car then limit what it can do.  If it had its own native dashcam app that could upload the videos when you get in wifi range of your home/office/whatever, that would be fantastic.  I plan on trying it out to see if it can handle it and not effect the main function (playing audio) but haven't got there yet, after 3d printing everything, multiple iterations of the hardware, I finally got it in the truck but now it's winter so playing with electronics outside isn't fun.
  Radio - minimum FM stereo, but since that requires an RTL-SDR to start with, why not incorporate at least a CB band radio but preferable much more such as weather radio, air band and for some users marine band.
  Always on top audio controls - no matter what app you're running there should always be at least large volume buttons on the screen and possibly next/previous track - not just when you have the player app on top.

  I don't want to seem like I'm bashing Openauto, it's awesome and mine is insanely loud and clear but I feel without full customization options it's a lot of work to end up with not much more than an android stereo at many times the price and complexity.  I can currently get an Android 10 stereo with a custom FJ Cruiser bezel for $132 CDN.  That's barely over the price of a Pi 7" screen ($120) let alone the Pi itself ($102 for a Pi4), a hifi hat ($76), an RTL-SDR dongle ($50) , all the components to make it work nice with vehicle voltages then amps ($198 for a 4 channel Pioneer) to drive the vehicle speakers.  Given that, it should be able to do WAY more than a standard Android stereo out of the gate.  Sure, a lot of it is the enjoyment of these projects but there comes a cost/benefit point.

  I'm probably not looking at this like most users, I'm more interested in having a fully self-sufficient solution geared more towards off-road/off-grid with no reliance on internet connectivity but still have the options for going fully connected when I get back.

  Willing to help anywhere I can.

**Reply by lamemakes** [reply]
Sorry for the delayed reply from me, but I really appreciate you taking the time to give all these details. This is super valuable and I would say you outlined my roadmap really well.

> Wallpaper - loved making my own wallpaper but since the control buttons take up so much real estate (and many buttons that I don't want and are not removable), I can't fully enjoy it. Would be nice to have the option of while on the main screen, the menu could push up to task bar and leave the current track playing across the bottom. And since I'm on wants.. animated wallpaper would rock! Even if just allowing animated gif support. I made a chromed "FJ Cruiser" logo in Blender, if it could just move ever so little the effect would be awesome.

This is interesting - I had considered a custom boot animation but the background would be something to explore.

> Native file browser - One of my reasons for wanting the Pi based solution is to get all my media into it and not rely on thumbdrives. There should be a simple way to copy from USB or a network share to internal storage. But with Openauto, as soon as I used the network for connecting to a phone (by accident as I can't remove the button), it overrode my network connection to my home wifi.

Totally agreed and this is a huge roadmap item. I'd like a lot more media support moving forward, and methods for the user to easily upload that. My biggest gripe with most headunits is that they're very reliant on separate media devices with little functionality (besides radio) on their own.

> Dashcam - why have this powerful little computer in your car then limit what it can do. If it had its own native dashcam app that could upload the videos when you get in wifi range of your home/office/whatever, that would be fantastic. I plan on trying it out to see if it can handle it and not effect the main function (playing audio) but haven't got there yet, after 3d printing everything, multiple iterations of the hardware, I finally got it in the truck but now it's winter so playing with electronics outside isn't fun.

This is something I hadn't full considered but would be cool to look ahead towards. I like the auto-export functionality.

> Radio - minimum FM stereo, but since that requires an RTL-SDR to start with, why not incorporate at least a CB band radio but preferable much more such as weather radio, air band and for some users marine band.

The dream! I completely agree. I'm a ham operator myself to to even be able to just passively listen would be super cool. Just not sure how much compute power that would require or maybe the focus needs to be on optimizing PILOT to make way for such processing power. Would also be pretty useful in some sort of emergency/off-grid scenario (this is just passive listening, transmit would be a different discussion haha).

> I'm probably not looking at this like most users, I'm more interested in having a fully self-sufficient solution geared more towards off-road/off-grid with no reliance on internet connectivity but still have the options for going fully connected when I get back.

This really was the main vision, so I totally agree. I think your vision certainly aligns with that of PILOT's! That being said, I recently tossed up a new announcement explaining that I have been and will continue to be taking a bit of a hiatus from active feature development. There's certainly a lot of technical debt involved with a lot of the features in my head and listed here, and being the sole developer sometimes stacks the motivation against me with a giant like this project. 

Development will continue in a few months, but in the meantime if you have any interest in digging into some Python feel free to take a stab at anything that might interest you! I'd love to outline some architecture of the project and anything else that may be useful. Thanks again for this fantastic feedback though! Hope we can see some of this implemented in the new year.
